### ISSUE TITLE
Reductions with an array of results gives memory error

### ISSUE THREAD

**Issue by Char-Aznable** [kokkos/kokkos#3727]
This code:
```c++
#include <limits>
#include <Kokkos_Core.hpp>

template < class ViewMatrix, class ViewVector >
struct ColumnwiseMax {

  using V = typename ViewMatrix::value_type;
  using size_type = typename ViewMatrix::size_type;

  using value_type = V[];

  KOKKOS_INLINE_FUNCTION
  void operator()(const size_type iRow, value_type update) const {
    for(size_type jCol = 0; jCol < value_count; ++jCol) {
      if(data_(iRow, jCol) > update[jCol]) {
        update[jCol] = data_(iRow, jCol);
      }
    }
  }

  KOKKOS_INLINE_FUNCTION void join(volatile value_type dst,
    const volatile value_type src) const {
    for(size_type i = 0; i < value_count; ++i) {
      if(src[i] > dst[i]) { dst[i] = src[i]; }
    }
  }

  KOKKOS_INLINE_FUNCTION void init(value_type dst) const {
    for(size_type i = 0; i < value_count; ++i) { dst[i] = -std::numeric_limits<V>::infinity(); }
  }

  using has_final = std::true_type;
  KOKKOS_INLINE_FUNCTION void final(value_type src) const {
    for(size_type i = 0; i < value_count; ++i) { results_(i) = src[i]; }
  }

  ViewMatrix data_;
  ViewVector results_;
  size_type value_count;

};

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv); {

  const size_t numRows = 4;
  const size_t numCols = 2374;
  using ViewMatrix = Kokkos::View<float**>;
  using ViewVector = Kokkos::View<float*>;

  ViewMatrix m ("m", numRows, numCols);
  ViewVector results("results", numCols);
  Kokkos::deep_copy(m, 1.0);
  // ... fill X before the following ...
  Kokkos::parallel_reduce (numRows,
    ColumnwiseMax<ViewMatrix,ViewVector>{m, results, numCols});

  Kokkos::fence();

  } Kokkos::finalize();
}

```
gives the error:
```c++
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaStreamSynchronize(m_stream) error( cudaErrorIllegalAddress): an illegal memory access was encountered /home/aznb/mycodes/testKokkos/build_cuda_debug/_deps/kokkos-src/core/src/Cuda/Kokkos_Cuda_Instance.cpp:314
Traceback functionality not available

```
maybe this is another memory alignment error?

**Comment by Char-Aznable** [comment]
BTW, the example follows the one on wiki: https://github.com/kokkos/kokkos/wiki/ParallelDispatch#733-reductions-with-an-array-of-results

**Comment by dalg24** [comment]
```diff
@@ -53,7 +53,7 @@
   Kokkos::deep_copy(m, 1.0);
   // ... fill X before the following ...
   Kokkos::parallel_reduce (numRows,
-    ColumnwiseMax<ViewMatrix,ViewVector>{m, results, numCols});
+    ColumnwiseMax<ViewMatrix,ViewVector>{m, results, numRows});

   Kokkos::fence();

**Comment by Char-Aznable** [comment]
@dalg24 What you proposed doesn't make sense if you read the rest of the program: it's supposed to do column-wise max so the result has to be of length `numCols`. In fact, setting `numCols` and `numRows` both to `2374` or any reasonably large number still triggers the bug.

**Comment by Char-Aznable** [comment]
On my maxwell52 GPU, it seems there is a hard upper limit on the value of `value_count` or `numCols`: any value above 361 triggers the bug but below 361 it doesn't

**Comment by dalg24** [comment]
> @dalg24 What you proposed doesn't make sense if you read the rest of the program: it's supposed to do column-wise max so the result has to be of length numCols. In fact, setting numCols and numRows both to 2374 or any reasonably large number still triggers the bug.

You're right.

> On my maxwell52 GPU, it seems there is a hard upper limit on the value of value_count or numCols: any value above 361 triggers the bug but below 361 it doesn't

I see the same thing on V100

**Comment by PhilMiller** [comment]
On 3.7.0, CUDA 11.2, debug build with bounds checking, a value of `numCols` as low as 550 results in throwing an exception:

> `Kokkos::Impl::ParallelReduce< Cuda > requested too much L0 scratch memory`

Backtrace is as follows:

```
#1  0x000000000040ed66 in Kokkos::Impl::throw_runtime_exception (msg=...)
#2  0x00000000004142d4 in Kokkos::Impl::check_reduced_view_shmem_size<void, Kokkos::RangePolicy<Kokkos::Cuda>, Test::ColumnwiseMax<Kokkos::View<float**>, Kokkos::View<float*> > > (functor=..., policy=...) at /net/watson.sandia.gov/storage/fast/projects/sems/install/rhel7-x86_64/sems-compilers/tpl/gcc/8.3.0/gcc/4.8.5/base/odjqidm/include/c++/8.3.0/bits/char_traits.h:352
#3  Kokkos::Impl::ParallelReduce<Test::ColumnwiseMax<Kokkos::View<float**>, Kokkos::View<float*> >, Kokkos::RangePolicy<Kokkos::Cuda>, Kokkos::InvalidType, Kokkos::Cuda>::ParallelReduce<Kokkos::View<float*, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> > > (arg_result=<synthetic pointer>..., arg_policy=..., arg_functor=..., this=0x7fffffffa0a0) at /ascldap/users/pbmille/repos/kokkos/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp:427
#4  Kokkos::Impl::ParallelReduceAdaptor<Kokkos::RangePolicy<Kokkos::Cuda>, Test::ColumnwiseMax<Kokkos::View<float**>, Kokkos::View<float*> >, Kokkos::View<float*, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> > >::execute_impl (return_value=<synthetic pointer>..., functor=..., policy=..., label=...) at /ascldap/users/pbmille/repos/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:1382
#5  Kokkos::Impl::ParallelReduceAdaptor<Kokkos::RangePolicy<Kokkos::Cuda>, Test::ColumnwiseMax<Kokkos::View<float**>, Kokkos::View<float*> >, Kokkos::View<float*, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> > >::execute<Kokkos::View<float*, Kokkos::HostSpace, Kokkos::MemoryTraits<1u> > > (return_value=<synthetic pointer>..., functor=..., policy=..., label=...) at /ascldap/users/pbmille/repos/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:1401
#6  Kokkos::parallel_reduce<Test::ColumnwiseMax<Kokkos::View<float**>, Kokkos::View<float*> > > (functor=..., policy=<synthetic pointer>: <optimized out>) at /ascldap/users/pbmille/repos/kokkos/core/src/Kokkos_Parallel_Reduce.hpp:1770
#7  Test::defaultdevicetype_development_test_Test::TestBody (this=<optimized out>) at /ascldap/users/pbmille/repos/kokkos/core/unit_test/default/TestDefaultDeviceDevelop.cpp:112
```

I feel like there was some more recent issue about this, too, but I can't immediately find it.

It would seem that `ParallelReduce` on CUDA is blindly trying to allocate scratch space sufficient for the array reduction, regardless of whether it will fit

**Comment by PhilMiller** [comment]
Christian is going to look at adding a host-side guard rail against this case, so it at least doesn't fail inscrutably.

**Comment by PhilMiller** [comment]
This seems to have been recapitulated in #4461, which was ostensibly addressed for CUDA in #4548. There's a follow-on in #4743.

**Comment by crtrott** [comment]
We at least don't crash anymore, but the error message is likely not super helpful. It says "requested too much L0 scratch". #4743 is the right place to track this.
